wdl-lint/src/rules/deprecated_placeholder_option.rs

```diff
@@ -0,0 +1,162 @@
+//! A lint rule for flagging placeholder options as deprecated.
+
+use wdl_ast::span_of;
+use wdl_ast::v1::Placeholder;
+use wdl_ast::v1::PlaceholderOption;
+use wdl_ast::version::V1;
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the deprecated placeholder option rule.
+const ID: &str = "DeprecatedPlaceholderOption";
+
+/// Creates a diagnostic for the use of the deprecated `default` placeholder
+/// option.
+fn deprecated_default_placeholder_option(span: Span) -> Diagnostic {
+    Diagnostic::warning(String::from(
+        "use of the deprecated `default` placeholder option",
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(
+        "replace the `default` placeholder option with a call to the `select_first()` standard \
+         library function",
+    )
+}
+
+/// Creates a diagnostic for the use of the deprecated `sep` placeholder option.
+fn deprecated_sep_placeholder_option(span: Span) -> Diagnostic {
+    Diagnostic::warning(String::from(
+        "use of the deprecated `sep` placeholder option",
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix(
+        "replace the `sep` placeholder option with a call to the `sep()` standard library function",
+    )
+}
+
+/// Creates a diagnostic for the use of the deprecated `true`/`false`
+/// placeholder option.
+fn deprecated_true_false_placeholder_option(span: Span) -> Diagnostic {
+    Diagnostic::warning(String::from(
+        "use of the deprecated `true`/`false` placeholder option",
+    ))
+    .with_rule(ID)
+    .with_highlight(span)
+    .with_fix("replace the `true`/`false` placeholder option with an `if`/`else` expression")
+}
+
+/// Detects the use of a deprecated placeholder option.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct DeprecatedPlaceholderOptionRule {
+    /// The detected version of the current document.
+    ///
+    /// The first [`Option`] represents whether the version has been hooked or
+    /// not by this rule.
+    ///
+    /// The second [`Option`] represents whether or not a valid version was able
+    /// to be parsed from the version string.
+    version: Option<Option<SupportedVersion>>,
```

Comment:
I'm not quite sure why this is `Option<Option<SupportedVersion>>`. Rules may expect that `document` is called before any other callbacks as it is the root of the walk, so should this not be `Option<SupportedVersion>`?

Reply:
You're correct that this assumption _should_ always remain true. The virtue that lead me to split into two `Option`s is simply to be able to handle the two different meanings of the previously overloaded `None` from the rule's perspective. Explicitly, collapsing the meanings and simply having the rule ignore things if the value is `None` is _okay_, but I felt it was a bit better to panic from the rule if the assumption it has is ever broken.

[`serde_with` uses a similar concept which I like quite a lot](https://docs.rs/serde_with/latest/serde_with/rust/double_option).

Though it introduces a small bit of complexity, it also makes things more explicit and provides an additional fail-safe if that assumption is every broken, so I'd argue to keep it (though I could be persuaded not to if the team felt strongly otherwise).

Reply:
I think by the nature of the CST having a root node and the root node being the "document", this invariant is very unlikely to ever be violated unless something is catastrophically wrong.

If we change this to `Option<SupportedVersion>`, then if somehow we violate the invariant, the rule won't fire a diagnostic rather than panicking like it does now with the double option; either way we'll observe the problem in our test coverage.

To me it just complicates the code (albeit it only slightly) and I certainly do a double take when I see a double option.

Reply:
Okay, I've finally found some time to dig into the details of the packages and how they work together, and now I have a better idea of the inner mechanisms here. But first, since this may come up as we continue to jointly develop out the package as a group, I should briefly share my perspective on the first part of your comment.

> I think by the nature of the CST having a root node and the root node being the "document", this invariant is very unlikely to ever be violated unless something is catastrophically wrong.
>
> If we change this to Option<SupportedVersion> and somehow we violate the invariant, the rule won't fire a diagnostic rather than panicking like it does now with the double option; either way we'll observe the problem in our test coverage.

To my eye, the difference in approach boils down to whether the benefit of being explicit (both in terms of the code as well as the error messages themselves) is worth the cost of introducing complexity or ugliness into the code.

As you say, it is quite unlikely that the assumption you laid out will ever be broken. However, in the case where it _is_ broken, I will always prefer to immediately get back an error telling me exactly what assumption was violated rather spending time tracking down why the majority of the tests are suddenly failing. Further, when there are two different reasons a `version` could be missing (in this case, because a catastrophic failure occurred in the parsing versus whether the version exists but failed to be parsed to a supporting version), I always prefer to make that immediately obvious and wrestle with those cases out in the open: the practical implications may be the same such that they could be collapsed, but the context with which I collapsed them is far more likely to be missed the next time I visit that code. You could encode those assumptions in comments, but I prefer for the code to yell at me when it can (😄).

As you might intuit from this perspective, I'm quite a defensive software engineer, as I've been bitten in big ways more than once by not taking this approach. However, I'm happy to be reigned in and accept the risks associated with not being explicit, as the likelihood that either approach will be consequential in the long term is small.

---

> Alternatively, we could also unwrap a Option<SupportedVersion> and continue to panic as visitors should not be receiving this callback on an unsupported document version (we skip parsing the contents of the file).

Now that I have a better understanding of what's going on underneath the hood, this comment spurred me to think that there is a better solution hidden behind a more fundamental question: do we want to allow `Visitor`s to traverse documents that aren't a supported WDL version?

To illustrate what happens today, take this document:

```
version 2.0

task {}
workflow {}
```

The root node and preamble are parsed, and the rest of the document is left as an unparsed node:

```
RootNode@0..33
  VersionStatementNode@0..11
    VersionKeyword@0..7 "version"
    Whitespace@7..8 " "
    Version@8..11 "2.0"
  Unparsed@11..33 "\n\ntask {}\nworkflow {}\n"
```

On its face, there is no reason not to allow this AST to be visited by the validator: documents with unsupported WDL versions produce an AST just the same as any supported versions. The big difference is that these ASTs only contain `RootNode` and the various version nodes, so only the `document()` and `version_statement()` callbacks will be executed.

That said, after giving it some deeper thought, I think there is very little benefit to allowing this behavior.

- I think there are very few pros to be had. Basically nothing can be confidently asserted about unsupported versions besides the fact that they are unsupported—any rules that do manage to execute could produce false positive or false negative results in a "silent" manner. Practically speaking, no callbacks besides `document()` and `version_statement_node()` would be run on this kind of AST, so none of the current rules will execute properly.
- The cost of allowing it means that every rule needs to now consider a greater surface area of whether the rule is valid for non-supported versions (a strange concept in general to me—the versions are unsupported by definition). This will lead to boilerplate that is needed in any version-dependent rule, which will become tiring.

Today, we almost always manually check if the parser produced any diagnostics before moving forward with validation/linting. For example, in the `wdl lint` command:

```
 async fn exec(self) -> Result<()> {
        let source = read_source(&self.path)?;
        let (document, diagnostics) = Document::parse(&source);

        // HERE, WE CHECK TO MAKE SURE NO DIAGNOSTICS OCCURRED WHILE PARSING
        // AND RETURN IF ANY ARE ENCOUNTERED.
        if !diagnostics.is_empty() {
            emit_diagnostics(&self.path.to_string_lossy(), &source, &diagnostics)?;

            bail!(
                "aborting due to previous {count} diagnostic{s}",
                count = diagnostics.len(),
                s = if diagnostics.len() == 1 { "" } else { "s" }
            );
        }

      
        let mut validator = Validator::default();
        validator.add_visitor(LintVisitor::default());
        if let Err(diagnostics) = validator.validate(&document) {
            emit_diagnostics(&self.path.to_string_lossy(), &source, &diagnostics)?;

            bail!(
                "aborting due to previous {count} diagnostic{s}",
                count = diagnostics.len(),
                s = if diagnostics.len() == 1 { "" } else { "s" }
            );
        }

        Ok(())
    }
```

There's only one place I spotted in the current implementation where we don't take this approach, and it's in [the engine code](https://github.com/stjude-rust-labs/wdl/blob/main/wdl-analysis/src/engine.rs#L552). When I was designing a solution that followed @peterhuene's suggestion to `unwrap()` the version, this was the cause of inconsistent behavior depending on what happened when I used `wdl lint` to test my solution versus when the engine took over during unit tests.

I'm still open to saying that we want to take the philosophy of allowing unsupported WDL documents to be visited, but, if we do that, we should probably go the full mile and implemented the `supported(version: Option<SupportedVersion>) -> bool` function to make this dance a little more graceful.

Reply:
Agreed, I think having `supported` on the rules / visitors makes a lot of sense.

I'm in support of resolving this issue as you see fit as I know completely understand why it's a double option 👍 .

Regarding the way that analysis is behaving, we could certainly make that consistent with everything else and only extend the diagnostics for a validation visit if parsing didn't have any documents.

Reply:
My preference is to roll with exactly what is implemented right now and merge it in so it makes the Sprocket Docket on Monday. My gut is that the `supported` callback is a bigger change, and this PR accomplishes enough such that we should kick that can down the road.

@a-frantz and @adthrasher, I plan to merge this Monday if there are no other objections, as Peter and I appear to land in the same place on this.